Step2 문자열 덧셈 계산기 PR 요청드립니다.

paki1019 opened the pull request:
Step2 문자열 계산기 PR 요청드립니다!

paki1019 committed:
test: 코드리뷰 반영

paki1019 committed:
feat : StringAddCalculator 테스트 코드 추가, 기능 구현

paki1019 committed:
refactor : StringAddCalculator

- 문자열 분리(커스텀 구분자) 메서드 분리
- number 원시 타입 포장, Number 클래스 생성

paki1019 committed:
refactor : StringAddCalculator

- 문자열 null, empty 체크 메서드 분리
- 상수 ZERO 접근제한자 public으로 변경
- numbers 일급 컬랙션 생성, Numbers 클래스 생성

paki1019 committed:
refactor : empty.txt 파일 제거, test 패키지 경로 파일, main 패키지로 이전

paki1019 committed:
refactor : 생성자 정적 패토리 메서드로 변경

paki1019 committed:
test : 빈 Numbers 합계 0 테스트 코드 추가

paki1019 committed:
refactor : 컨벤션 적용.

seokhongkim commented on src/main/java/stringaddcalculator/Number.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Objects;
+
+public class Number {
```
값 객체 사용 좋습니다. 👍

seokhongkim commented on src/main/java/stringaddcalculator/Number.java:
```diff
@@ -0,0 +1,47 @@
+		return new Number(n);
+	}
+
+	public boolean isNegative() {
```
도메인 요구사항이 숫자는 양수여야 한다는 것이니, `Number` 객체는 음수로 생성할 수 없어야 할 것 같아요.
그렇다면 생성자에서 밸리데이션을 해보는 것은 어떨까요?

paki1019 replied:
넵, 생성자에서 음수 여부를 검증하는 `validateNotNegative()` 로직 추가하였습니다! 기존의 `isNegative()`, `conatinsNegative()` 메서드도 제거 되어서 확실히 코드가 더 간결해진것 같습니다!

seokhongkim commented on src/main/java/stringaddcalculator/Number.java:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	public Number sum(Number y) {
+		return new Number(this.number + y.number);
```
불변 객체로 유지하신 것 좋습니다. 👍

seokhongkim commented on src/main/java/stringaddcalculator/Numbers.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Numbers {
```
일급 컬렉션 구현 좋습니다. 👍

seokhongkim commented on src/main/java/stringaddcalculator/Numbers.java:
```diff
@@ -0,0 +1,31 @@
+	public Number sum() {
+		return numbers.stream()
+			.reduce(Number::sum)
+			.orElse(Number.from(Number.ZERO));
```
`Number.ZERO` 는 내부에서 0이라는 기준 값으로 사용되는 것으로 보이고, 이 코드의 맥락에서는 값이 없을 때 `기본 반환값` 의 의미로 사용되는 것 같습니다.
동일한 값이지만 다른 의미를 가지므로, 별도의 상수로 관리해보는 것도 괜찮다고 생각되는데, 진호님은 어떻게 생각하시나요?

paki1019 replied:
넵, 말씀하신대로 같은 값이여도 맥락이 다르면 별도의 상수로 관리하는것이 맞을것 같습니다.
(ZERO 상수를 public으로 선언해야할 때부터 약간 냄새가 났던것 같습니다...)

Numbers 클래스의 0 값을 `SUM_DEFAULT_RETURN_VALUE`로,
StringAddCalculator 클래스의 0 값을 `NULL_OR_EMPTY_DEFAULT_RETURN_VALUE`로 수정하겠습니다!

seokhongkim commented on src/main/java/stringaddcalculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
+	private static final String DEFAULT_SPLIT_REGEX = ",|:";
```
상수로 패턴을 관리하시는 것 좋습니다. 👍

seokhongkim commented on src/main/java/stringaddcalculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,44 @@
+		return result.getValue();
+	}
+
+	private static String[] splitText(String text) {
```
`Numbers` 를 통해 계산하는 책임을 잘 나눠주셨으니, 문자열을 파싱하는 것도 `InputSplitter` 와 같은 클래스를 이용해서 역할을 나눠보면 어떨까요?

paki1019 replied:
`InputSplitter` 클래스로 분리하였습니다. 확실히 역할을 더 나누니, 코드가 더 간결해지고,`InputSplitter`에 대한 테스트 코드 짜는것도 더 편해진 것 같습니다!

seokhongkim commented on src/test/java/stringaddcalculator/NumberTest.java:
```diff
@@ -0,0 +1,22 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class NumberTest {
```
꼼꼼한 단위 테스트 작성 좋습니다. 👍

seokhongkim commented on src/test/java/stringaddcalculator/NumberTest.java:
```diff
@@ -0,0 +1,22 @@
+		assertThat(Number.from(3).isNegative()).isFalse();
+		assertThat(Number.from(-1).isNegative()).isTrue();
+	}
+}
```
파일 마지막에 newline 이 없다고 경고를 띄워주네요.
파일 마지막의 newline 은 왜 필요할까요?

paki1019 replied:
검색해본 결과, 이는 POSIX 표준으로, 파일 끝에 개행을 추가함으로써 텍스트 파일간 구분을 더 용이하게 하기 위한 목적이라고 합니다.
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

지식이 늘었습니다👍

seokhongkim reviewed (changes requested):
진호님, 2단계 미션 잘 구현해주셨습니다. 👍 

`Number`, `Numbers` 를 이용해 숫자를 관리하는 역할을 분리하신 것 좋습니다. 👍 
기능 목록을 `README.md` 파일에 작성하신 후 테스트를 작성하는 과정으로 진행해보시면 학습에 도움이 될 것 같습니다. :)

조금 더 역할을 나눠보면 좋을 것 같은 부분이 있어 코멘트 남겼습니다.
코멘트 확인 부탁드립니다. 🙏 

코멘트 보시고 궁금하신 점은 깃헙 코멘트나 슬랙 DM으로 편하게 말씀 부탁드립니다. 😀

paki1019 committed:
refactor : 코드 리뷰 반영 - Number 값 객체 생성자에서 음수 밸리데이션

paki1019 committed:
refactor : 코드 리뷰 반영 - 파일 마지막에 누락된 newline 추가

paki1019 committed:
docs : README.md step 2 내용 추가

paki1019 committed:
test: 숫자 이외의 값 전달 테스트 케이스 추가

paki1019 committed:
refactor: 코드 리뷰 반영 - 상수값 분리

paki1019 committed:
refactor: 코드 리뷰 반영 - 문자열 파싱 로직 InputSplitter 클래스로 분리

paki1019 committed:
fix: Number 숫자가 아닌 문자열 전달시 throw 되는 예외 타입 수정

- validate  음수값 전달, 숫자 이외의 값 전달 테스트 코드 수정
- NumberFormatException 예외 RuntimeException으로 예외 번역

seokhongkim reviewed (approved):
진호님, 2단계 피드백 잘 반영해주셨습니다! 💯 

작업을 작게 나눠서 진행하시고 커밋도 잘게 나눠주셔서 코드 파악에 더 도움이 되었습니다. 👍 
2단계 PR은 이만 머지하도록 하겠습니다.

3단계 미션에서 뵙겠습니다. 😀
